contracts/PangolinExchange.sol

```diff
@@ -0,0 +1,166 @@
+  IPangolinRouter pangolinRouter;
+  SupportedAssets supportedAssets;
+
+  /* ========= CONSTRUCTOR ========= */
+
+  constructor (address _pangolinRouter, SupportedAssets _supportedAssets) {
+    pangolinRouter = IPangolinRouter(_pangolinRouter);
+    supportedAssets = _supportedAssets;
+  }
+
+  /* ========= MODIFIERS ========= */
+
+  modifier RefundRemainder {
+    _;
+    if (address(this).balance > 0) {
+      (bool success,) = msg.sender.call{value : address(this).balance}("");
+      require(success, "Refund failed");
+    }
+  }
+
+
+  function transferBack(bytes32 _asset) external override {
+    address tokenAddress = supportedAssets.getAssetAddress(_asset);
+    IERC20 token = IERC20(tokenAddress);
+    token.transfer(msg.sender, token.balanceOf(address(this)));
+  }
+
+
+  /**
+   * Buys selected ERC20 token with AVAX using the Pangolin DEX
+   * Refunds unused AVAX to the msg.sender
+   * @dev _token ERC20 token's address
+   * @dev _exactERC20AmountOut amount of the ERC20 token to be bought
+  **/
+  function buyAsset(bytes32 _token, uint256 _exactERC20AmountOut) payable external override RefundRemainder {
+    require(_exactERC20AmountOut > 0, "Amount of tokens to buy has to be greater than 0");
+
+    address tokenAddress = supportedAssets.getAssetAddress(_token);
+
+    uint256 amountIn = getEstimatedAVAXForERC20Token(_exactERC20AmountOut, tokenAddress);
+
+    require(msg.value >= amountIn, "Not enough funds provided");
+
+    pangolinRouter.swapAVAXForExactTokens{value : msg.value}(_exactERC20AmountOut, getPathForAVAXtoToken(tokenAddress), msg.sender, block.timestamp);
+
+    emit TokenPurchase(msg.sender, _exactERC20AmountOut, block.timestamp);
+  }
+
+
+  /**
+   * Sells selected ERC20 token for AVAX
+   * @dev _token ERC20 token's address
+   * @dev _exactERC20AmountIn amount of the ERC20 token to be sold
+   * @dev _minAvaxAmountOut minimum amount of the AVAX token to be bought
+  **/
+  function sellAsset(bytes32 _token, uint256 _exactERC20AmountIn, uint256 _minAvaxAmountOut) external override RefundRemainder {
+    require(_exactERC20AmountIn > 0, "Amount of tokens to sell has to be greater than 0");
+    address tokenAddress = supportedAssets.getAssetAddress(_token);
+
+    IERC20 token = IERC20(tokenAddress);
+    token.approve(address(pangolinRouter), _exactERC20AmountIn);
+    pangolinRouter.swapExactTokensForAVAX(_exactERC20AmountIn, _minAvaxAmountOut, getPathForTokenToAVAX(tokenAddress), msg.sender, block.timestamp);
+
+    emit TokenSell(msg.sender, _exactERC20AmountIn, block.timestamp);
+  }
+
+  /* ========== RECEIVE AVAX FUNCTION ========== */
+  receive() external payable {  }
+
+
+  /* ========== VIEW FUNCTIONS ========== */
+
+  /**
+    * Returns the current balance of the asset held by a given user
+    * @dev _asset the code of an asset
+    * @dev _user the address of queried user
+  **/
+  function getBalance(address _user, bytes32 _asset) external override view returns(uint256) {
```

Comment:
Do we need this method here, it's not connected to the exchange?

Reply:
That's a good point. It is defined in `IAssetsExchange` and implemented in this contract but actually, it is only being used in the `SmartLoan.sol` itself. I will refactor this.

Reply:
Done